Parallelized inner loop over sets in MinHash generation with user-defined reduction

compiling-is-winning opened the pull request:
See title.  Also for diagnosing bug in interactive test case where one of the verification arrays has a value changed by calling lshMinMax()

Ethan-DeBandi99 commented:
@compiling-is-winning a few things that would be helpful for us to debug.

- Can you create a README.md somewhere that lists the dependencies for this module? It looks like you are pulling in some C code. Want to make sure I install the correct thing to test.
- In test can you add a script that just runs through and provides expected vs actual output. I know you sent us over your workflow, but have the script that we can easily run would be super helpful. Don't worry about assertions, we can validate just by output.

compiling-is-winning commented:
Is this fine for the readme file:

https://github.com/compiling-is-winning/arkouda-contrib/tree/main/arkouda_cws_hashing

and likewise this for a test script:

https://github.com/compiling-is-winning/arkouda-contrib/blob/main/arkouda_cws_hashing/test/lsh_minmax.py

I added the latter awhile ago, and the attached workflow document is basically just that test run in the terminal (because I can't ever recall how to run the test script).

Also, I switched two of the inner forall loops of `lshMinMax()` to use the "may-parallel" version of the forall, since it is probably often better to allow them to run in serial anyway.  Do either of you know (or have any pointers to documentation) on how the "may-parallel" construct works?  E.g. is it like Cilk or OpenMP tasks, complete with work-stealing and all (doubt it)?  Does it at least calculate the size of the loop to determine if parallelization is beneifical?  Or does it just assign threads on a first-come, first-served basis, so that, say, a loop with 10 elements might be parallelized while one with 10^6 might be run in serial?
